Documentation/design-docs/arm64-intrinsics.md

```diff
@@ -0,0 +1,217 @@
+## Arm64 Intrinsics
+
+This document is intended to document proposed design decision related to the introduction
+of Arm64 Intrinsics
+
+### Intrinsics in general
+
+Use of intrinsics in general is a CoreCLR design decision to allow low level platform
+specific optimizations.  
+
+At first glance, such a decision seems to violate the fundamental principles of .NET
+code running on any platform.  However, the intent is not for the vast majority of
+apps to use such optimizations.  The intended usage model is to allow library
+developers access to low level functions which enable optimization of key
+functions.  As such the use is expected to be limited, but performance critical.
+
+#### Intrinsic `IsSupported` granularity
+
+The existing implementation is using `IsSupported` to determine whether an existing
+namespace is supported on a given platform.
+
+The granularity is kept at a large scale to have large blocks of functionality controlled
+by a single `IsSupported`.
+
+#### Intrinsic granularity
+
+In general individual intrinsic will be chosen to be fine grained.  These will generally correspond to a single assembly instruction.
+
+#### Intrinsics & Crossgen
+
+For any intrinsic which may not be supported on all variants of a platform.  Crossgen
+Method compilation must be trapped, so that the JIT is forced to generate optimal platform dependent 
+code at runtime.
+
+#### API review
+
+Intrinsics will extend the API of CoreCLR.  They will need to follow standard API review practices.
+
+### Choice of Arm64 naming conventions
+
+#### Namespaces
+
+`ARM32` and `ARM64` will follow similar namespace conventions.
+
++ `System.Runtime.Intrinsics` is used for type definitions useful across multiple platforms
++ `System.Runtime.Intrinsics.Arm` is used for enumerations shared across `ARM32` and `ARM64` platforms
++ `System.Runtime.Intrinsics.Arm.Base` is recommended for functionality included in all `ARM32` and `ARM64` platforms
+  + Do we need this or could we assume `ARMv8` is minimum supported platform?
+  + Should we assume `ARMv7` or `ARMv5` is minimum supported platform?
++ `System.Runtime.Intrinsics.Arm.ARMv8` is used for `ARMv8` added functionality included in both `ARM32` and `ARM64` platforms 
++ `System.Runtime.Intrinsics.Arm.ARMv8.Arm64` is used for `ARMv8` functionality only included in `ARM64` platform
++ `System.Runtime.Intrinsics.Arm.ARMv8_1....` is used for `ARMv8.1` functionality.  
```

Comment:
IMO there should be separate namespace for SVE intrinsics. I do not see a reason to split Arm intrinsics between ARMv8_1, ARMv8_2 unless there are new ISA extensions available. How about using Neon namespace for old style intrinsics?

Reply:
ARMv8.1 and ARMv8.2 are separate extensions, they are each separate and are not required to be implemented together.    I have not refreshed my memory of ISA extensions added by each architectural extension.  It may be that ARMv8.2 implements no new instructions.  In which case, the class would not be introduced.

Reply:
`Neon` is not a term used in ARM64 terminology.  It is a legacy term introduced in the ARMv7 timeframe.  I do not think its use would aid comprehension.

Reply:
I think it might. The latest version of the `ARM Architecture Reference Manual` still has the following except:
> The Advanced SIMD architecture, its associated implementations, and supporting software, are commonly referred to as NEON™ technology.

Anyone using the ARM intrinsics will likely be familiar with the term `NEON`

Reply:
It looks like `NEON` is exclusively used for the `AArch32` while `Advanced SIMD` is used exclusively for `AArch64`, in the actual manual. Linux seems to disambiguate these via HW_NEON and HW_ASIMD capability bits.

https://github.com/torvalds/linux/blob/master/arch/arm/include/uapi/asm/hwcap.h
https://github.com/torvalds/linux/blob/master/arch/arm64/include/uapi/asm/hwcap.h